File: challenge-340/deadmarshal/haskell/ch1.hs
module Ch1 where

import Data.List

duplicateRemovals :: String -> String
duplicateRemovals = go []
  where
    go stack [] = reverse stack
    go [] (x:xs) = go [x] xs
    go (s:stack) (x:xs)
      | s == x = go stack xs
      | otherwise = go (x:s:stack) xs
    
main :: IO ()
main = do
  print $ duplicateRemovals "abbaca"
  print $ duplicateRemovals "azxxzy"
  print $ duplicateRemovals "aaaaaaaa"
  print $ duplicateRemovals "aabccba"
  print $ duplicateRemovals "abcddcba"

File: challenge-340/deadmarshal/java/Ch1.java
public class Ch1 {
  public static void main(String[] args) {
    System.out.println(duplicate_removals("abbaca"));
    System.out.println(duplicate_removals("azxxzy"));
    System.out.println(duplicate_removals("aaaaaaaa"));
    System.out.println(duplicate_removals("aabccba"));
    System.out.println(duplicate_removals("abcddcba"));
  }

  private static String duplicate_removals(String str) {
    StringBuilder sb = new StringBuilder();
    for (char c : str.toCharArray()) {
      if ((sb.length() > 0) && (sb.charAt(sb.length() - 1) == c))
        sb.deleteCharAt(sb.length() - 1);
      else
        sb.append(c);
    }
    return sb.toString();
  }
}
